python/cuda_parallel/cuda/parallel/experimental/iterators/_iterators.py

```diff
@@ -337,11 +336,55 @@ def dereference(state):
             return op(it_dereference(state))
 
         def __hash__(self):
-            return hash((self._it, self._op))
+            return hash((self.__class__.iterator_kind_type, self._it, self._op))
 
         def __eq__(self, other):
             if not isinstance(other.kind, TransformIteratorKind):
                 return NotImplemented
             return self._it == other._it and self._op == other._op
 
     return TransformIterator(it, op)
+
+
+def make_advanced_iterator(it: IteratorBase, /, *, offset: int = 1):
+    it_advance = cuda.jit(type(it).advance, device=True)
+    it_dereference = cuda.jit(type(it).dereference, device=True)
+
+    class AdvancedIteratorKind(IteratorKind):
+        pass
```

Comment:
I don't know if this matters at all, but the existing code has this outside the function.

For self-consistency, I usually follow the existing pattern. (If something is wrong one way or another, the failures are less surprising if there is a common style.)

Reply:
It does matter. The kind type determines abi suffix appended to method name, i.e, `advance_xxxxxxxx`. 

For advanced iterator generated internally we needs this suffix to change every time iterator's ``__add__`` special method is called.

Otherwise the program below will fail to link due to multiple symbol definitions.

```python
it = ...
it1 = it + 1
it2 = it + 2
func(it1, it2)
...
```